core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -196,15 +212,15 @@ public SingleDistributionTaskSource(ListMultimap<PlanNodeId, ExchangeSourceHandl
         }
 
         @Override
-        public List<TaskDescriptor> getMoreTasks()
+        public ListenableFuture<List<TaskDescriptor>> getMoreTasks()
         {
             List<TaskDescriptor> result = ImmutableList.of(new TaskDescriptor(
                     0,
                     ImmutableListMultimap.of(),
                     exchangeSourceHandles,
                     new NodeRequirements(Optional.empty(), ImmutableSet.of(), taskMemory)));
             finished = true;
-            return result;
+            return immediateFuture(result);
```

Comment:
This is a pre-existing problem, but i wonder if you could add a

```
if(finished){
  return immediateFuture(ImmutableList.of());
}
```

check here and in the `ArbitraryDistributionTaskSource` while you are at it? (maybe as a separate commit)?